trustymail/trustymail.py

```diff
@@ -415,15 +406,67 @@ def scan(domain_name, timeout, smtp_timeout, smtp_localhost, smtp_ports, smtp_ca
     return domain
 
 
-def handle_error(prefix, domain, error):
-    if hasattr(error, "message"):
-        if "NXDOMAIN" in error.message and prefix != "[DMARC]":
+def handle_error(prefix, domain, error, syntax_error=False):
+    """Handle an error by logging via the Python logging library and
+    recording it in the debug_info or syntax_error members of the
+    trustymail.Domain object.
+
+    Since the "Debug Info" and "Syntax Error" fields in the CSV output
+    of trustymail come directly from the debug_info and syntax_error
+    members of the trustymail.Domain object, and that CSV is likely
+    all we will have to reconstruct how trustymail reached the
+    conclusions it did, it is vital to record as much helpful
+    information as possible.
+
+    Parameters
+    ----------
+    prefix : str
+        The prefix to use when constructing the log string.  This is
+        usually the type of trustymail test that was being performed
+        when the error condition occurred.
+
+    domain : trustymail.Domain
+        The Domain object in which the error or syntax error should be
+        recorded.
+
+    error : str, BaseException, or Exception
+        Either a string describing the error, or an exception object
+        representing the error.
+
+    syntax_error : bool
+        If True then the error will be recorded in the syntax_error
+        member of the trustymail.Domain object.  Otherwise it is
+        recorded in the error member of the trustymail.Domain object.
+    """
+    # Get the previous frame in the stack - the one that is calling
+    # this function
+    frame = inspect.currentframe().f_back
+    function = frame.f_code
+    function_name = function.co_name
+    filename = function.co_filename
+    line = frame.f_lineno
+
+    error_template = '{prefix} In {function_name} at {filename}:{line}: {error}'
+
+    if hasattr(error, 'message'):
+        if syntax_error and 'NXDOMAIN' in error.message and prefix != '[DMARC]':
             domain.is_live = False
-        domain.errors.append(error.message)
-        logging.debug("  {0} {1}".format(prefix, error.message))
+        error_string = error_template.format(prefix=prefix, function_name=function_name, line=line, filename=filename, error=error.message)
     else:
-        domain.errors.append(str(error))
-        logging.debug("  {0} {1}".format(prefix, str(error)))
+        error_string = error_template.format(prefix=prefix, function_name=function_name, line=line, filename=filename, error=str(error))
+
+    if syntax_error:
+        domain.syntax_errors.append(error_string)
+    else:
+        domain.debug_info.append(error_string)
+    logging.debug(error_string)
+
+
+def handle_syntax_error(prefix, domain, error):
+    """Convenience method for handle_error(prefix, domain, error,
```

Comment:
@jsf9k -- Minor nit, maybe we should just say "... for `handle_error(...)`", to shorten the length and to work around any function signature changes later.